pkg/extensions/extensions.go

```diff
@@ -148,41 +150,79 @@ func (h *Host) RegisterInprocExtension(name string, ext apiv1.Implementation) {
 }
 
 // getExtensionPath returns the path to an extension binary
-func (h *Host) getExtensionPath(version, name string) string {
+func (h *Host) getExtensionPath(commit, name string) string {
+	fmt.Printf("commit: %v\n", commit)
 	homeDir, _ := os.UserHomeDir() //nolint:errcheck // Why: signature doesn't allow it, yet
-	path := filepath.Join(homeDir, ".outreach", ".cache", "stencil", "extensions", name, fmt.Sprintf("@%s", version), filepath.Base(name))
+	path := filepath.Join(homeDir, ".outreach", ".cache", "stencil", "extensions", name, fmt.Sprintf("@%s", commit), filepath.Base(name))
 	os.MkdirAll(filepath.Dir(path), 0o755) //nolint:errcheck // Why: signature doesn't allow it, yet
 	return path
 }
 
+// getVersionWithCommit retrieves a new version with the commit present.
+func getVersionWithCommit(ctx context.Context, token cfg.SecretData, repoURL string,
+	version *resolver.Version) (*resolver.Version, error) {
+	var v *resolver.Version
+	var err error
+	// We assume that if the tag does not comply with semver it is a channel (e.g. unstable).
+	if _, err := semver.ParseTolerant(version.Tag); err != nil {
+		v, err = resolver.Resolve(ctx, token, &resolver.Criteria{
+			URL:     repoURL,
+			Channel: version.Tag,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest version")
+		}
+		return v, nil
+
+	}
+	v, err = resolver.Resolve(ctx, token, &resolver.Criteria{
+		URL:         repoURL,
+		Constraints: []string{version.Tag},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get latest version")
+	}
+	return v, nil
+}
+
 // downloadFromRemote downloads a release from github and extracts it to disk
 //
 // using the example extension module: github.com/getoutreach/stencil-plugin
 //
 //	org: getoutreach
 //	repo: stencil-plugin
 //	name: github.com/getoutreach/stencil-plugin
-func (h *Host) downloadFromRemote(ctx context.Context, name, version string) (string, error) {
+func (h *Host) downloadFromRemote(ctx context.Context, name string,
+	version *resolver.Version) (string, error) {
 	token, err := github.GetToken()
 	if err != nil {
 		h.log.WithError(err).Warn("Failed to get github token, falling back to anonymous")
 	}
 
 	repoURL := "https://" + name
 
-	if version == "" {
+	if version.Tag == "" {
 		v, err := resolver.Resolve(ctx, token, &resolver.Criteria{
 			URL: repoURL,
 		})
 		if err != nil {
 			return "", errors.Wrap(err, "failed to get latest version")
 		}
-		version = v.Tag
+		version = v
+	}
+
+	if version.Commit == "" {
+		v, err := getVersionWithCommit(ctx, token, repoURL, version)
+		if err != nil {
+			return "", errors.Wrap(err, "retrieving commit")
+
+		}
+		version = v
 	}
 
 	// Check if the version we're pulling already exists and is executable before downloading
 	// it again.
-	dlPath := h.getExtensionPath(version, name)
+	dlPath := h.getExtensionPath(version.Commit, name)
```

Comment:
I'd prefer to not lose the struct here, I'd recommend continuing to pass the version and handle the logic in `getExtensionPath` instead. It should be easier to test that way as well